specs/merge/fork-choice.md

```diff
@@ -66,44 +67,59 @@ def finalize_block(self: ExecutionEngine, block_hash: Hash32) -> bool:
     ...
 ```
 
-## Containers
+## Helpers
 
-#### `PowBlock`
+### `TransitionStore`
 
 ```python
-class PowBlock(Container):
+@dataclass
+class TransitionStore(object):
+    transition_total_difficulty: uint256
+```
+
+### `PowBlock`
+
+```python
+@dataclass
+class PowBlock(object):
     block_hash: Hash32
     is_processed: boolean
     is_valid: boolean
     total_difficulty: uint256
 ```
 
-## Helper functions
-
-#### `get_pow_block`
+### `get_pow_block`
 
 Let `get_pow_block(block_hash: Hash32) -> PowBlock` be the function that given the hash of the PoW block returns its data.
 
 *Note*: The `eth_getBlockByHash` JSON-RPC method does not distinguish invalid blocks from blocks that haven't been processed yet. Either extending this existing method or implementing a new one is required.
 
-#### `is_valid_transition_block`
+### `get_transition_store`
+
+```python
+def get_transition_store(anchor_pow_block: PowBlock):
+    transition_total_difficulty = anchor_pow_block.total_difficulty + TRANSITION_TOTAL_DIFFICULTY_OFFSET
```

Comment:
Shouldn't we just make this:

```suggestion
    transition_total_difficulty = anchor_pow_block.total_difficulty + anchor_pow_block.difficulty * (86400 * 7 // 13)
```

This way we have to deal with less time uncertainty because the difficulty would be set at the anchor time instead of weeks before it.

Reply:
In normal case there are difficulty fluctuations caused by stochastic nature of PoW and hashpower ins/outs and if we take only one block to estimate this large period the estimation might become not that accurate. In our case this block is at least `2048` block behind current head and doesn't give us representation on what's happening in the network right at the moment of computation; theoretically, real hashrate could be doubled during the last `2048` blocks. If anchor block was produced in a period of 51% attack then we can get its difficulty halved wrt the normal case difficulty (another half of hashpower used by attacker or honest miners).

All the inputs above make me think that we better take a constant that to our observation will worth 1 week of hashpower contributed to the network in a normal case. We can take a large period of time to estimate it and get rid of poisson and hashpower fluctuations. But if there is a drop in hashrate right before the Merge estimated offset will become inaccurate which will increase time to transition.

Both approaches can give an unpredictable result depending on the network conditions. But the latter one is more secure and conservative from my point of view because real moment of transition can happen not earlier than the half of expected period (in case if hashpower magically doubles). And this is why I am leaning towards it.

We also need to take the last voting period in account to compute the offset, 
i.e. `anchor_pow_block.difficulty * (2048 + 2048 * 12 // 13 + 86400 * 7 // 13)`.

Reply:
> In our case this block is at least 2048 block behind current head and doesn't give us representation on what's happening in the network right at the moment of computation

Right, but any difficulty offset determined statically at fork client deployment time (weeks before fork time) would be _even less_ accurate. So it seems better to just make a formula that calculates the offset in real time taking the anchor pow block difficulty into consideration. If we want to prevent miners from attacking by coordinating to turn off their miners for a day at fork time, we could also take the maximum of the difficulty at fork client deployment time and the `anchor_pow_block.difficulty`.

> We also need to take the last voting period in account to compute the offset,
i.e. anchor_pow_block.difficulty * (2048 + 2048 * 12 // 13 + 86400 * 7 // 13).

This makes sense!

Reply:
This is rather a trade-off between more accurate and more secure. Statically determined offset (or a block difficulty) could be safer than the one requested from the network, it depending on the network conditions of course.

> we could also take the maximum of the difficulty at fork client deployment time and the `anchor_pow_block.difficulty`

What if we do the following: `anchor_difficulty = max(MIN_ANCHOR_DIFFICULTY, anchor_pow_block.difficulty)`, and define the minimal safe difficulty manually cause it, again, should be safer than deferring it to a short moment of network's lifetime?